### ISSUE TITLE
[New]: Make BEAMZ designs fully differentiable.

### ISSUE THREAD

**Issue by QuentinWach** [QuentinWach/beamz#44]
https://x.com/QuentinWach/status/2008317445098902006

I think I will make BEAMZ fully differentiable.
Let's kill large parameter sweeps.
Just define your objective and the geom. constraints.

---

Here is ChatGPT giving some example ideas and explaining why this is important:

## 1) Parameterized geometry optimization (engineers actually do this)

**Problem:** Optimize a *few meaningful CAD parameters* (width, gap, bend radius), not millions of density pixels.

**Why differentiable matters:** You want `d(loss)/d(width)` directly, not `d(loss)/dρ` and then hand-wiring your own chain rule every time.

```python
import beamz as bz

# Parameters (these are "trainable")
w = bz.Parameter(500e-9)        # waveguide width
g = bz.Parameter(200e-9)        # coupler gap
L = bz.Parameter(20e-6)         # coupling length

# Geometry defined from parameters
geom = bz.geometry.DirectionalCoupler(width=w, gap=g, length=L)

# Materials, sources, monitors
si = bz.materials.Si(n=3.48)
sio2 = bz.materials.SiO2(n=1.44)

sim = bz.Simulation(
    geometry=geom,
    materials={"core": si, "clad": sio2},
    wavelength=1550e-9,
    bc="PML",
)

# Objective: maximize power in desired port, minimize in the other
def loss_fn():
    fields = sim.run()
    p_through = bz.monitor.power(fields, port="through")
    p_cross   = bz.monitor.power(fields, port="cross")
    return -(p_cross - 0.1 * p_through)  # negative = maximize p_cross

# One call gives gradients wrt CAD parameters
loss = loss_fn()
grads = bz.grad(loss, wrt=[w, g, L])      # or loss.backward()

bz.optim.Adam(lr=1e-2).step([w, g, L], grads)
```

**What this unlocks:** your users can do *real photonic design* (with meaningful params) using the same workflow they use in ML.

---

## 2) Fabrication-aware inverse design (the thing papers gloss over)

**Problem:** Designs fail after fabrication because of min-feature, etch bias, corner rounding.

**Differentiable trick:** Treat fabrication as differentiable operators: blur (litho), erosion/dilation (etch bias approx), projection (binarization).

```python
rho = bz.ParameterField(shape=(256, 256), init="random")  # continuous density

def fabricate(rho):
    # Differentiable approximations
    rho = bz.ops.gaussian_blur(rho, sigma=1.2)          # litho resolution
    rho = bz.ops.bias(rho, amount=-0.02)                # etch bias (approx)
    rho = bz.ops.sigmoid_projection(rho, beta=20.0)     # push to 0/1
    return rho

def loss_fn():
    rho_fab = fabricate(rho)
    geom = bz.geometry.FromDensity(rho_fab, pixel_size=20e-9)
    sim.geometry = geom

    fields = sim.run()
    T = bz.monitor.transmission(fields, "out")
    R = bz.monitor.reflection(fields, "in")
    return -(T - 0.2*R)

loss = loss_fn()
rho_grad = bz.grad(loss, wrt=[rho])[0]
bz.optim.Adam(lr=5e-2).step([rho], [rho_grad])
```

**Real-world value:** “works in simulation” becomes “works after tapeout”.

---

## 3) Robust design (worst-case / expected performance)

**Problem:** Device must work across wavelength, temperature, and manufacturing variations.

**Differentiable value:** You want gradients of an expectation / worst-case metric, *not* hand-tuned sweeps.

```python
rho = bz.ParameterField((256, 256), init="seeded")

wavelengths = bz.linspace(1500e-9, 1600e-9, 11)
temps       = [290, 300, 310]  # K
etch_biases = [-10e-9, 0.0, +10e-9]

def loss_fn():
    losses = []
    for lam in wavelengths:
        for T in temps:
            for b in etch_biases:
                rho_fab = bz.ops.fabrication_stack(rho, etch_bias=b)
                n_core  = bz.models.si_index(wavelength=lam, temperature=T)

                sim.set(wavelength=lam, materials={"core": bz.materials.Index(n_core)})
                sim.geometry = bz.geometry.FromDensity(rho_fab)

                fields = sim.run()
                losses.append(-bz.monitor.transmission(fields, "out"))

    # Option A: expected loss
    return bz.mean(bz.stack(losses))

    # Option B: worst-case loss
    # return bz.max(bz.stack(losses))

loss = loss_fn()
grad = bz.grad(loss, wrt=[rho])[0]
bz.optim.Adam(lr=2e-2).step([rho], [grad])
```

**Why this matters commercially:** Industry cares about yield; academia cares about “it works once”.

---

## 4) EM + Thermal co-optimization (first real “multiphysics” step)

**Problem:** You design an optical heater/tuner. EM defines absorption → heating. Thermal changes refractive index → changes EM performance.

**Differentiable value:** End-to-end gradient across coupled solvers.

```python
heater = bz.ParameterField((256, 256), init="zeros")   # where to put metal/heater
rho    = bz.ParameterField((256, 256), init="seeded")  # photonic structure

def loss_fn():
    # --- EM solve (forward) ---
    geom = bz.geometry.Composite([
        bz.geometry.FromDensity(rho),
        bz.geometry.HeaterFromDensity(heater),
    ])
    sim.geometry = geom

    fields = sim.run()
    absorption = bz.physics.em_absorption(fields, region="heater")  # W/m^3 map

    # --- Thermal solve ---
    temp = bz.thermal.solve(
        heat_source=absorption,
        k_map=bz.materials.thermal_conductivity_map(geom),
        bc="fixed_edges",
    )

    # --- Thermo-optic coupling back into EM ---
    dn = bz.physics.thermo_optic_dn(temp, material="Si")
    sim.materials = bz.materials.apply_dn(sim.materials, dn)

    fields2 = sim.run()
    T = bz.monitor.transmission(fields2, "out")

    # Penalize heater power and maximize transmission
    P = bz.sum(absorption)  # proxy for electrical/absorbed power
    return -(T - 1e-3 * P)

loss = loss_fn()
grads = bz.grad(loss, wrt=[rho, heater])
bz.optim.Adam(lr=1e-2).step([rho, heater], grads)
```

**This is the bridge** from “photonic EM code” → “multiphysics platform”.

---

## 5) Gradient-based design of experiments (inverse problems)

**Problem:** You measure a spectrum and want to infer geometry/material parameters (metrology / tomography).

**Differentiable value:** You solve it like a fitting problem with gradients.

```python
# Unknowns to infer from measurement
w = bz.Parameter(480e-9)
h = bz.Parameter(220e-9)
n = bz.Parameter(3.45)

measured = load_spectrum("measurement.csv")

def loss_fn():
    geom = bz.geometry.Waveguide(width=w, height=h)
    sim.geometry  = geom
    sim.materials = {"core": bz.materials.Index(n), "clad": bz.materials.SiO2()}

    pred = []
    for lam in measured.wavelength:
        sim.set(wavelength=lam)
        fields = sim.run()
        pred.append(bz.monitor.transmission(fields, "out"))

    pred = bz.stack(pred)
    return bz.mse(pred, measured.transmission)

loss = loss_fn()
grads = bz.grad(loss, wrt=[w, h, n])
bz.optim.LBFGS().step([w, h, n], grads)
```

**This is “physics as a differentiable model”** — it’s the same pattern as training a neural net, but the model is Maxwell + your geometry.

---

# The architectural pattern behind all of these

You’ll notice every example wants the same thing:

1. **Parameters** (scalars or fields)
2. **Transforms** (fabrication, smoothing, projection, coupling)
3. **Physics ops** (EM solve, thermal solve, etc.)
4. **Measurements** (power, phase, spectrum)
5. **Loss**
6. **grad()** + optimizer

That’s the PyTorch-like “physical programming” loop.

---

# Minimal pseudo-API you need to support this cleanly

This is the “make it real” sketch:

```python
class Parameter:
    value: float
    grad: float

class ParameterField:
    value: array
    grad: array

def grad(loss, wrt): ...
def vjp(physics_op, primals, cotangent): ...
```

And crucially: your EM solve should be a **custom differentiable primitive** whose VJP is implemented by your adjoint method (not by tracing timesteps).

Conceptually:

```python
fields = em_solve(geometry, materials, sources)      # forward

# VJP uses your adjoint solver:
dL_dgeom, dL_dmat = em_solve.vjp(dL_dfields)
```

That is what makes gradients *composable*.

**Comment by QuentinWach** [comment]
Backlog audit (2026-08-06): BeamZ now has differentiable design transforms, but the FDTD solve itself is not yet a composable differentiable primitive.

Already available:

- JAX/Optax density filtering, morphology, Heaviside/SSP1 projection, continuation, fixed-structure masks, and VJP mapping through the transform pipeline.
- Manual forward/adjoint overlap accumulation and a disk-backed field-history utility.
- One-step time-reversibility and low-level directional-derivative validation for existing primitives.

Core work still required:

- [ ] Move trainable material values into dynamic runtime arguments so optimization iterations reuse one compiled executable; values must not enter the compiled-program cache key.
- [ ] Define a custom VJP for the complete FDTD solve rather than exposing naive reverse-mode tracing through every timestep.
- [ ] Implement checkpointed/rematerialized forward-state recovery with a bounded memory budget. Reversible gradients can follow only where loss, PML, sources, and monitor state have a mathematically valid inverse treatment.
- [ ] Generate adjoint sources for field, flux, modal/S-parameter, and later far-field objectives.
- [ ] Accumulate gradients at the correct Yee/material supports and propagate them through raster/filter/projection transforms.
- [ ] Support multiple frequencies, sources, process corners, and batch/sharding strategies without recompilation per scalar parameter value.
- [ ] Validate 2D/3D gradients against finite differences under refinement, including CPML, tensor smoothing, continuation, and modal normalization.
- [ ] Define the first stable public API around explicit parameter pytrees and `value_and_grad`; do not require mutable `Parameter.grad` state.

Parameterized CAD differentiation should follow the material-grid primitive through a differentiable SDF/boundary representation; the Rust rasterizer should not be presented as differentiable until a valid VJP is defined. #162 tracks a later broadband storage optimization, #27 tracks concurrent scheduling, and #75 tracks strict fabrication constraints.
